File: python/617.py
"""Project Euler Problem 617: Mirror Power Sequence.

Find the number of (n, e)-MPS sequences defined by a_{i+1} = min(a_i^e, n - a_i^e),
a_i > 1, and n ≤ N.

Given a "tower power" sequence (b, b^e, b^{e²}, ... b^{e^k}), there are two
types of MPS sequences:
- A cyclic permutation of the tower powers of b (k+1 of these)
- A cyclic permutation of a suffix of the sequence (k of these).

We iterate over all such tower power sequences, and check for each of the above
types of MPS sequences whether n ≤ N.
"""

from __future__ import annotations

from math import isqrt


def pow_int(base: int, exp: int) -> int:
    """Integer power."""
    result = 1
    for _ in range(exp):
        result *= base
    return result


def solve() -> int:
    """Solve Problem 617."""
    N = 10**18

    ans = isqrt(N) - 2
    a0 = 2
    while pow_int(a0, 3) + a0 <= N:
        e = 2
        while pow_int(a0, e) + a0 <= N:
            as_list = []
            a = a0
            while pow_int(a, e) <= N:
                as_list.append(a)
                a = pow_int(a, e)

            for start in range(len(as_list)):
                for end in range(start, len(as_list)):
                    if (e > 2 or end > 0) and pow_int(as_list[end], e) + as_list[start] <= N:
                        ans += end + 1 if start == 0 else 1
            e += 1
        a0 += 1

    return ans


def main() -> int:
    """Main entry point."""
    result = solve()
    print(result)
    return result


if __name__ == "__main__":
    main()

File: c/617.c
/*
 * Project Euler Problem 617: Mirror Power Sequence
 *
 * Count (n, e)-MPS sequences where n <= 10^18.
 */
#include <stdio.h>
#include <stdint.h>
#include <math.h>

typedef long long ll;
typedef unsigned long long ull;
typedef __int128 i128;

static ll isqrt_ll(ll n) {
    ll x = (ll)sqrtl((long double)n);
    while (x * x > n) x--;
    while ((x + 1) * (x + 1) <= n) x++;
    return x;
}

/* Safe integer power: returns 0 if overflow would occur past limit */
static ll safe_pow(ll base, int exp, ll limit) {
    ll result = 1;
    for (int i = 0; i < exp; i++) {
        if (result > limit / base) return limit + 1; /* overflow sentinel */
        result *= base;
    }
    return result;
}

int main(void) {
    ll N = 1000000000000000000LL; /* 10^18 */

    ll ans = isqrt_ll(N) - 2;

    for (ll a0 = 2; ; a0++) {
        ll a0_cubed = safe_pow(a0, 3, N);
        if (a0_cubed + a0 > N) break;

        for (int e = 2; ; e++) {
            ll a0_e = safe_pow(a0, e, N);
            if (a0_e + a0 > N) break;

            /* Build tower */
            ll as_list[100];
            int as_len = 0;
            ll a = a0;
            while (1) {
                ll ae = safe_pow(a, e, N);
                if (ae > N) break;
                as_list[as_len++] = a;
                a = ae;
                if (as_len >= 100) break;
            }

            for (int start = 0; start < as_len; start++) {
                for (int end = start; end < as_len; end++) {
                    ll ae = safe_pow(as_list[end], e, N);
                    if (ae > N) break;
                    if ((e > 2 || end > 0) && ae + as_list[start] <= N) {
                        ans += (start == 0) ? end + 1 : 1;
                    }
                }
            }
        }
    }

    printf("%lld\n", ans);
    return 0;
}
